sklearn/preprocessing/_encoders.py

```diff
@@ -36,8 +55,19 @@ def _check_X(self, X):
           constructed feature by feature to preserve the data types
           of pandas DataFrame columns, as otherwise information is lost
           and cannot be used, eg for the `categories_` attribute.
-
+          If categories == 'dtypes' and the pandas column is a category,
+          the pandas series will be returned in this list.
         """
+        if self.categories == 'dtypes':
+            X_dtypes = getattr(X, "dtypes", None)
+            if hasattr(self, "_X_fit_dtypes"):  # fitted
+                if not self._check_dtypes_equal(self._X_fit_dtypes,
+                                                X_dtypes):
+                    raise ValueError("X.dtypes must match the dtypes used "
```

Comment:
Why is this check needed in general? Don't we already check categories on transform? 
I think we should be careful here, dtypes in pandas can change for other reasons like presence of missing values or not, and int/float mixture seems to work right now (eg fitting with int, transforming with float)

Reply:
Or maybe only doing it for categorical dtype?

Reply:
It's not in the future, but now: if you have a dataset with integers, but for some reason you have a missing values in part of your dataset, they become float. Even after dropping NaNs (because OneHotEncoder does not yet support that), they stay float. So it is not unreasonable to read your train and test data from separate files and end up with a different dtype for that reason. 

Note that we still check that the categories are found, or no new ones are there, etc.

Reply:
> Note that we still check that the categories are found, or no new ones are there, etc.

This is what the PR does now.

> Even after dropping NaNs (because OneHotEncoder does not yet support that), they stay float. 

Okay lets keep this float/int support.